### PR TITLE
Remove boost strings split usage

### LINKED ISSUE #662: Building without depending on libboost?
Hi there, @lethal-guitar !

On my continuing mission for native engines to try on SDL2/KMSDRM (SDL2 with no X11 depencencies!), I have just found this engine, and looks & sounds awesome on videos.

So I have tried to build it but seems to depend on boost, which is a HUGE lib!
Is there a way to build without depending on it? Checked CMakeLists.txt, but couldn't find a thing about it...

And, since it seems to run on OGA machines, I believe it must me possible to build without the heavy BOOST dependencies.

Thanks!

Manuel

EDIT: Even after installing libboost-dev, I still get this with `cmake -DUSE_GL_ES=ON -DCMAKE_BUILD_TYPE=Release ..`

```
manuel@hp15db0:~/src/RigelEngine/b4$ cmake -DUSE_GL_ES=ON -DCMAKE_BUILD_TYPE=Release ..
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/Boost-1.71.0/BoostConfig.cmake:117 (find_package):
  Could not find a package configuration file provided by
  "boost_program_options" (requested version 1.71.0) with any of the
  following names:

    boost_program_optionsConfig.cmake
    boost_program_options-config.cmake

  Add the installation prefix of "boost_program_options" to CMAKE_PREFIX_PATH
  or set "boost_program_options_DIR" to a directory containing one of the
  above files.  If "boost_program_options" provides a separate development
  package or SDK, be sure it has been installed.
Call Stack (most recent call first):
  /usr/lib/x86_64-linux-gnu/cmake/Boost-1.71.0/BoostConfig.cmake:182 (boost_find_component)
  /usr/share/cmake-3.16/Modules/FindBoost.cmake:443 (find_package)
  CMakeLists.txt:53 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/manuel/src/RigelEngine/b4/CMakeFiles/CMakeOutput.log".
```

### PR DESCRIPTION
**Pull request by bencsikandrei**
Hello,

This one resolves #662, partially.
I've started with string `split` since I knew how to implement that.
Funny bit is that it ended up a bit faster than boost's version..

You should check out the benchmarks. I guess that's understandable since this version I added is much simpler and only uses a simple delimiter (one char).

But hey, I'll take it.

EDIT: add `trim[Right|Left]` and `startsWith`

Cheers!

### CODE REVIEW

**Comment by lethal-guitar** [comment]
Nice one, really appreciate it! 👍🏻 I'll check it out as soon as I can 🙂

File: src/common/string_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace rigel::strings {
+
+/** Split a string based on a single char delimiter and put the result into output
+ */
+void split(std::vector<std::string> &output, std::string const &input,
```
Comment (lethal-guitar) [inline src/common/string_utils.hpp:26]: I'd prefer returning the result, instead of using an out parameter. That would deviate more from the Boost API but I don't think that's an issue. Since there are not that many call sites it's easy to review the replacement.

Not sure why the Boost API was designed that way, maybe they didn't trust RVO back in the day? Or maybe it's for when you'd like to split multiple input strings into a single destination container? But that's not really a use case we have in RigelEngine right now, so we can go with a more straight-forward function signature.

File: test/test_string_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    CHECK(v.size() == 2);
+    CHECK(v.front() == "hello");
+    CHECK(v.back() == " world");
+  }
+  SECTION("String with other white space delimiters") {
+    std::vector<std::string> v;
+    rigel::strings::split(v, "hello\nworld", '\n');
+    CHECK(v.size() == 2);
+  }
+}
```
Comment (lethal-guitar) [inline test/test_string_utils.cpp:47]: Another edge case that could be worth testing is a string that ends with a delimiter, like `"hello\n".

File: test/test_string_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  }
+  SECTION("String with no delimiter") {
+    std::vector<std::string> v;
+    rigel::strings::split(v, "hello", ' ');
+    CHECK(v.size() == 1);
+  }
+  SECTION("String with delimiter") {
+    std::vector<std::string> v;
+    rigel::strings::split(v, "hello, world", ',');
+    CHECK(v.size() == 2);
```
Comment (lethal-guitar) [inline test/test_string_utils.cpp:38]: Suggestion for making the test assertions a bit nicer:

```cpp
const auto expected = std::vector<std::string>{"hello", " world"};
CHECK(v == expectedResult);
```

That way, both the length of the result as well as the values are checked in one statement.

You could also use deduction guides and literal postfix to make it even shorter:

```cpp
using namespace std::string_literals;
const auto expected = std::vector{"hello"s, " world"s};
```

Up to you which one you prefer, both work for me 🙂

File: src/common/string_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace rigel::strings {
```
Comment (lethal-guitar) [inline src/common/string_utils.hpp:22]: I'd prefer to have this file in `src/base` instead of `src/common`. The former is where I mostly have general purpose utilities/standard library extensions, and string utilities fit in there quite well.

`common` on the other hand is more of a place for things that are needed to glue `frontend` and the rest of the code together - at least that's the idea, but in practice it mostly ended up being the place where I put things that don't really fit anywhere else 😅
 
Admittedly, that's not very clear from the name of the directory, I should think about a better name at some point..

File: src/common/string_utils.cpp [hunk 5]
```diff
@@ -0,0 +1,19 @@
+#include "string_utils.hpp"
+
+#include <algorithm>
+
+namespace rigel::strings {
+
+void split(std::vector<std::string> &output, std::string const &input,
+           char delimiter) {
+  auto start = std::begin(input);
```
Comment (lethal-guitar) [inline src/common/string_utils.cpp:9]: I wonder if we should `assert` that the delimiter is a plain ASCII character, i.e. less than 127? That way, string splitting can work correctly for UTF-8 encoded strings, without any surprises (as long as the delimiter is ASCII).

RigelEngine doesn't use Unicode anywhere so far, so it's not really a concern nowadays. But in the more distant future, I could imagine adding some localization features as an enhancement over the original game, so it would be a bit of future proofing.

What do you think?

**Review by lethal-guitar — COMMENTED** [review]
Thanks a lot for helping out with the effort to get rid of Boost dependencies! And thanks for the comprehensive PR, I appreciate the tests & benchmarks and the corresponding README additions. 👍🏻 Nice to hear that the new version is faster. As you say, it's maybe not too surprising, with the Boost version being much more general.

I left a few comments regarding some details, but this already looks great overall, thanks again!

File: src/common/string_utils.hpp (hunk 1 above)
Comment (bencsikandrei) [reply src/common/string_utils.hpp:26]: I think they had to work with a million compilers, some of which we might have never even heard about. And those might not do RVO properly.
So happy you want to use RVO. From now on I'll definitely do that first. The only reason I did it like that was for the compatibility-ish..
Will do

File: test/test_string_utils.cpp (hunk 3 above)
Comment (bencsikandrei) [reply test/test_string_utils.cpp:38]: Sure thing, Have to admit I've not used Catch in a while!
CTAD is a thing of beauty, but somehow I prefer the first one (call me old fashion), and if both are OK with you I'll go with it.

File: src/common/string_utils.cpp (hunk 5 above)
Comment (bencsikandrei) [reply src/common/string_utils.cpp:9]: Will do.
Looks like something that can save some trouble in the long run, so I say why not.
Will be added along with the other changes.

**Comment by bencsikandrei** [comment]
@lethal-guitar I've started looking at ways to replace 'startsWith' and 'trim' (and its flavors).

how do you want to proceed? do we do it in this same MR, or want me to open others for each of those? I'd really like to use the benchmarks here to see how much my changes differ from the boost:: functionality.

but hey, I can find a way

**Comment by lethal-guitar** [comment]
@bencsikandrei cool! 👍🏻 I'd say feel free to add more commits to this PR, then we have all the Boost-related changes in one place

File: src/base/string_utils.cpp [hunk 6]
```diff
@@ -20,4 +20,9 @@ std::vector<std::string> split(std::string const &input, char delimiter) {
   return output;
 }
 
+bool startsWith(std::string_view input, std::string_view prefix)
+{
+  return input.size() >= prefix.size() && input.substr(0, prefix.size()) == prefix;
```
Comment (lethal-guitar) [inline src/base/string_utils.cpp:25]: This is simple and easy to read 👍🏻 so I would say it's fine if it's a little slower compared to the Boost version. We don't use it that frequently. And once we move to C++ 20, we can switch to the one from the standard lib (https://en.cppreference.com/w/cpp/string/basic_string/starts_with).

Comment (bencsikandrei) [reply src/base/string_utils.cpp]: You gave me an awesome idea! Since C++20 is already partially supported by gcc/clang, I looked at how an implementation of this looks like and I made a small change that ended up being about 30% faster than the previous one, whilst remaining  readable (and extremely fast in debug mode - almost the same as in release actually).

Comment (lethal-guitar) [reply src/base/string_utils.cpp]: Nice one! 👍🏻

Comment (lethal-guitar) [reply src/base/string_utils.cpp]: Just out of curiosity, I had a quick look (https://github.com/boostorg/algorithm/blob/c9077bd49580bddfaa8a5c71abc4022a2b09d972/include/boost/algorithm/string/predicate.hpp) but I'm not really sure what makes it faster 🤔 it's more or less the same, except that they check both iterators against their respective range end on each iteration, instead of comparing the sizes upfront like you do. I would expect the latter to be faster since there should be fewer instructions per loop iteration, but who knows.. Could be interesting to compare the generated assembly.

In any case, your new version looks great to me 👍🏻

**Comment by bencsikandrei** [comment]
If my search was correct that should be all of them. The only boost requirement you still have is the one you're fixing in #665

**Comment by lethal-guitar** [comment]
> If my search was correct that should be all of them.

Awesome, I'll have a look!

File: benchmark/bench_string_utils.cpp [hunk 7]
```diff
@@ -97,3 +97,42 @@ static void BMBoostStartsWithFalseCase(benchmark::State &state) {
 }
 
 BENCHMARK(BMBoostStartsWithFalseCase);
+
+static void BMTrimLeft(benchmark::State &state) {
+  std::string kInputString = "  1234 ";
```
Comment (lethal-guitar) [inline benchmark/bench_string_utils.cpp:102]: I think you need to construct the string inside the benchmark loop, since the trimming mutates it in place - so after the first iteration, each subsequent trim call won't need to trim anything anymore.

We are going to also measure the string construction/destruction in that case, but assuming small string optimization, the time for that should be constant, so the comparison with the Boost implementation would still be valid I believe.

Comment (bencsikandrei) [reply benchmark/bench_string_utils.cpp:102]: dang, how did I miss that?!

fixed now, plus, used benchmark's API to not time the string creation (dtor may still be timed tho)

**Comment by lethal-guitar** [comment]
For future reference, benchmark results on my Windows Desktop:

```
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 256 KiB (x8)
  L3 Unified 12288 KiB (x1)
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BMStringSplit                         126 ns          126 ns      4977778
BMBoostStringSplit                    718 ns          698 ns       896000
BMStartsWithTrueCase                 7.83 ns         7.95 ns    112000000
BMStartsWithTrueCaseLongString       8.69 ns         8.72 ns     89600000
BMStartsWithFalseCase                7.77 ns         7.85 ns     89600000
BMBoostStartsWithTrueCase            3.13 ns         3.15 ns    203636364
BMBoostStartsWithFalseCase           1.44 ns         1.44 ns    497777778
BMTrimLeft                            482 ns          547 ns      1000000
BMTrimRight                           480 ns          570 ns      1672533
BMTrim                                497 ns          594 ns      1000000
BMBoostTrim                           674 ns          767 ns       896000
```

**Comment by lethal-guitar** [comment]
Hey @bencsikandrei, thanks again for this PR! I finally had the time to properly test everything, and it works great as far as I can tell. I made a few minor changes:

* Split adding the benchmarks into a dedicated commit
(more on why in a sec)
* Made `startsWith` `noexcept` as well - even though `std::equal` isn't marked as such, it should never throw in practice so that seemed fine to me. Let me know if you think that doesn't make sense!
* Made the trim unit tests a little more terse by inlining the expected value into the `CHECK` macro - this works fine for something like string literals, but looks a bit odd when doing it with `vector` or similar, so I usually prefer to have it outlined in a variable for the latter.
* Changed the CMake option name to `RIGEL_BUILD_BENCHMARKS`, as plural seemed a bit more fitting and it's also consistent with the option description that way.

Now this is basically ready to be merged from my side, except there's one little problem with the benchmarks.. I'm building with `-Werror` by default and on Mac/clang, that causes the build of googlebenchmark to fail due to warnings. I normally use `rigel_disable_warnings(target_name)` for 3rd party code, but it wasn't clear to me how I would apply that to the googlebenchmark target.

Maybe it would be better to not define `-Werror` globally, and instead have a `rigel_enable_warnings_as_errors()` helper function in CMake, and then apply that to all Rigel targets, instead of disabling warnings for 3rd party code?

Another thought I had was that we should probably build (even if we don't run them) the benchmarks on CI, as otherwise it might be too easy to accidentally break them for one of the platforms?

Let me know what you think, I'd be happy to merge this now without the benchmarks and add them back in a separate PR, or we can also sort it out as part of this PR - as you prefer 🙂

**Comment by bencsikandrei** [comment]
Hey @lethal-guitar!

> * Split adding the benchmarks into a dedicated commit
>   (more on why in a sec)

Awesome work splitting it up. It's cleaner this way.

> * Made `startsWith` `noexcept` as well - even though `std::equal` isn't marked as such, it should never throw in practice so that seemed fine to me. Let me know if you think that doesn't make sense!
Good choice, my fingers were also itching to type that but I resisted because of equal not being noexcept. Glad you pulled the trigger. Even if it would throw, I think `std::terminate` is a good choice there, not like we're handling the exception anyway.

> * Changed the CMake option name to `RIGEL_BUILD_BENCHMARKS`, as plural seemed a bit more fitting and it's also consistent with the option description that way.
I agree!

> * Removed the reference to #662 from the commit message, since the issue won't be fully resolved until #665 has been merged.
Wasn't sure about this either, but good choice.


> Now this is basically ready to be merged from my side, except there's one little problem with the benchmarks.. I'm building with `-Werror` by default and on Mac/clang, that causes the build of googlebenchmark to fail due to warnings. I normally use `rigel_disable_warnings(target_name)` for 3rd party code, but it wasn't clear to me how I would apply that to the googlebenchmark target.

I think we can hack our way around this by setting some options on the `benchmark` target (maybe you already tried that?), but I  think I have a better way of handling it (it involves something that you wrote a few lines after).

> Maybe it would be better to not define `-Werror` globally, and instead have a `rigel_enable_warnings_as_errors()` helper function in CMake, and then apply that to all Rigel targets, instead of disabling warnings for 3rd party code? Then again, it would also be annoying to see warnings while the 3rd party targets are built.. 
^^ This.

> 
> Another thought I had was that we should probably build (even if we don't run them) the benchmarks on CI, as otherwise it might be too easy to accidentally break them for one of the platforms?

Agree.

> Let me know what you think, I'd be happy to merge this now without the benchmarks and add them back in a separate PR, or we can also sort it out as part of this PR - as you prefer
I think I can cook some cmake-foo rather quickly. I'll then activate the benchmarks in CI to see if it build.

File: CMakeLists.txt [hunk 8]
```diff
@@ -18,7 +18,7 @@ endif()
 
 option(USE_GL_ES "Use OpenGL ES instead of regular OpenGL" OFF)
 option(WARNINGS_AS_ERRORS "Treat compiler warnings as errors" ON)
-
+option(BUILD_BENCHMARKS "Build benchmarks" OFF)
```
Comment (bencsikandrei) [inline CMakeLists.txt:21]: Slight issue here, the flags don't agree. One's called BUILD_BENCHMARKS the other at line 98 checks for RIGEL_BUILD_BENCHMARKS

File: benchmark/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    GIT_REPOSITORY https://github.com/google/benchmark.git
+    GIT_TAG        v1.5.3
+)
+FetchContent_GetProperties(benchmark)
+if(NOT benchmark_POPULATED)
+    set(BENCHMARK_ENABLE_GTEST_TESTS OFF)
+    set(BENCHMARK_ENABLE_INSTALL OFF)
+    set(BENCHMARK_ENABLE_TESTING OFF)
+    FetchContent_Populate(benchmark)
+    add_subdirectory(${benchmark_SOURCE_DIR} ${benchmark_BINARY_DIR})
```
Comment (bencsikandrei) [inline benchmark/CMakeLists.txt:14]: we could do:
```
    # disable warnings for both targets inside the google `benchmark` build
    rigel_disable_warnings(benchmark)
    rigel_disable_warnings(benchmark_main)
```
but I would steer clear of this one since it's quite brittle. 
do we care if third party code has warnings when it builds itself? I would say no.
the better option is to have granular compiler warnings and set it ONLY on RigelEngine code, not on third party stuff.

File: CMakeLists.txt (hunk 8 above)
Comment (lethal-guitar) [reply CMakeLists.txt:21]: Ah sorry, forgot to change the other one!

File: benchmark/CMakeLists.txt (hunk 9 above)
Comment (lethal-guitar) [reply benchmark/CMakeLists.txt:14]: Yeah that sounds good and cleaner/easier to maintain 👍🏻

Comment (bencsikandrei) [reply benchmark/CMakeLists.txt:14]: These changes are present in the latest commit.
I'll resolve this since I started it and the changes are there.

Comment (lethal-guitar) [reply benchmark/CMakeLists.txt:14]: Sure thing! 👍🏻 Thanks for making the changes!

**Comment by lethal-guitar** [comment]
More benchmark results (2014 MacBook Pro):

```
Run on (4 X 2800 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x2)
  L1 Instruction 32 KiB (x2)
  L2 Unified 256 KiB (x2)
  L3 Unified 3072 KiB (x1)
Load Average: 18.30, 11.13, 9.31
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BMStringSplit                         219 ns          218 ns      3202591
BMBoostStringSplit                   2337 ns         2344 ns       300370
BMStartsWithTrueCase                 3.98 ns         3.98 ns    174966756
BMStartsWithTrueCaseLongString       15.4 ns         15.3 ns     45158960
BMStartsWithFalseCase                2.94 ns         2.94 ns    236419394
BMBoostStartsWithTrueCase            2.47 ns         2.47 ns    281142563
BMBoostStartsWithFalseCase           1.22 ns         1.22 ns    564752961
BMTrimLeft                           1486 ns         1480 ns       471196
BMTrimRight                          1487 ns         1484 ns       468519
BMTrim                               1501 ns         1500 ns       465766
BMBoostTrim                          1592 ns         1590 ns       431398
```

**Comment by lethal-guitar** [comment]
Aaaand it's merged. 🎉  Thanks again for this really nice contribution!!
